File: duplicate-charge-redirect-webhook-race/node/refund-duplicates.js
/**
 * Find WooCommerce orders charged twice on Stripe and refund the extra charge.
 * Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/duplicate-charge-redirect-webhook-race/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 48);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function recentCharges(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  const out = [];
  for await (const ch of stripe.charges.list({ limit: 100, created: { gte: since } })) out.push(ch);
  return out;
}

export function duplicateSets(charges) {
  const byOrder = new Map();
  for (const ch of charges) {
    const oid = (ch.metadata || {}).order_id;
    if (oid && ch.status === "succeeded" && !ch.refunded) {
      if (!byOrder.has(oid)) byOrder.set(oid, []);
      byOrder.get(oid).push(ch);
    }
  }
  const duplicates = new Map();
  for (const [oid, group] of byOrder) {
    const byAmount = new Map();
    for (const ch of group) {
      if (!byAmount.has(ch.amount)) byAmount.set(ch.amount, []);
      byAmount.get(ch.amount).push(ch);
    }
    for (const [amount, same] of byAmount) {
      if (same.length > 1) duplicates.set(`${oid}:${amount}`, { oid, same });
    }
  }
  return duplicates;
}

export function chooseExtras(same, orderTransactionId) {
  let keeper = same.find((c) => c.id === orderTransactionId);
  if (!keeper) keeper = same.reduce((a, b) => (a.created <= b.created ? a : b));
  return same.filter((c) => c.id !== keeper.id);
}

export async function run() {
  const charges = await recentCharges(LOOKBACK_HOURS);
  const duplicates = duplicateSets(charges);
  let refunded = 0;
  for (const { oid, same } of duplicates.values()) {
    const order = await woo(`/orders/${oid}`);
    const extras = chooseExtras(same, order ? order.transaction_id : null);
    for (const charge of extras) {
      console.log(`Order ${oid}: duplicate charge ${charge.id}. ${DRY_RUN ? "would refund" : "refunding"}`);
      if (!DRY_RUN) {
        await stripe.refunds.create({ charge: charge.id, reason: "duplicate" });
        await woo(`/orders/${oid}/notes`, {
          method: "POST",
          body: JSON.stringify({
            note: `Refunded duplicate Stripe charge ${charge.id} ` +
                  `(${charge.amount} ${charge.currency}). Kept the charge on the order.`,
          }),
        });
      }
      refunded++;
    }
  }
  console.log(`Done. ${refunded} duplicate charge(s) ${DRY_RUN ? "to refund" : "refunded"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: duplicate-charge-redirect-webhook-race/python/refund_duplicates.py
"""Find WooCommerce orders charged twice on Stripe and refund the extra charge.
Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/woocommerce/duplicate-charge-redirect-webhook-race/
"""
import os
import time
import logging
from collections import defaultdict
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("refund_duplicates")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "48"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def recent_charges(lookback_hours):
    since = int(time.time()) - lookback_hours * 3600
    return stripe.Charge.list(limit=100, created={"gte": since}).auto_paging_iter()


def duplicate_sets(charges):
    by_order = defaultdict(list)
    for ch in charges:
        oid = (ch.get("metadata") or {}).get("order_id")
        if oid and ch.get("status") == "succeeded" and not ch.get("refunded"):
            by_order[oid].append(ch)
    duplicates = {}
    for oid, group in by_order.items():
        by_amount = defaultdict(list)
        for ch in group:
            by_amount[ch["amount"]].append(ch)
        for amount, same in by_amount.items():
            if len(same) > 1:
                duplicates[(oid, amount)] = same
    return duplicates


def order_transaction_id(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json().get("transaction_id")


def choose_extras(same, order_transaction):
    keeper = next((c for c in same if c["id"] == order_transaction), None)
    if keeper is None:
        keeper = min(same, key=lambda c: c["created"])
    return [c for c in same if c["id"] != keeper["id"]]


def add_note(order_id, note):
    requests.post(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
                  json={"note": note}, auth=AUTH, timeout=30).raise_for_status()


def run():
    charges = list(recent_charges(LOOKBACK_HOURS))
    duplicates = duplicate_sets(charges)
    refunded = 0
    for (order_id, amount), same in duplicates.items():
        extras = choose_extras(same, order_transaction_id(order_id))
        for charge in extras:
            log.info("Order %s: duplicate charge %s for %s. %s",
                     order_id, charge["id"], amount, "would refund" if DRY_RUN else "refunding")
            if not DRY_RUN:
                stripe.Refund.create(charge=charge["id"], reason="duplicate")
                add_note(order_id, f"Refunded duplicate Stripe charge {charge['id']} "
                                   f"({charge['amount']} {charge['currency']}). Kept the charge on the order.")
            refunded += 1
    log.info("Done. %d duplicate charge(s) %s.", refunded, "to refund" if DRY_RUN else "refunded")


if __name__ == "__main__":
    run()
